### PR TITLE
Added new features and made some modifications to function within our infrastructure

### PR DESCRIPTION
**Pull request by tr-eliu**
Added features:
Added a brute forcing rule to the waf
Added cloudwatch alarms to all the rules to the waf, and have them generate alerts on pagerduty via sns
Configure waf to send logs into Splunk
Modifications:
added variables to make it easier to change names of certain resources
remove the bash script in place to run the code and instead added a main.tf
Added tags to certain resources

### CODE REVIEW

File: wafwebacl.tf [hunk 1]
```diff
@@ -77,4 +77,44 @@ resource "aws_waf_web_acl" "WAFWebACL" {
...
+
+    provisioner "local-exec" {
+        command = "aws cloudwatch put-metric-data --namespace WAF --metric-name BadBotRule --unit Count --value 0 --dimensions Rule=${aws_waf_rule.WAFBadBotRule.metric_name},WebACL=${aws_waf_web_acl.WAFWebACL.metric_name}"
+    }
+
+    provisioner "local-exec" {
+        command = "aws cloudwatch put-metric-data --namespace WAF --metric-name SQLInjectionRule --unit Count --value 0 --dimensions Rule=${aws_waf_rule.WAFSqlInjectionRule.metric_name},WebACL=${aws_waf_web_acl.WAFWebACL.metric_name}"
+    }
+
+    provisioner "local-exec" {
```
Comment (svaranasi-traderev) [inline wafwebacl.tf]: Sorry, why are we having to use local-exec? I think we talked about this, but I no longer remember the discussion.

File: WAFBrokenAuthSessionManageRule.tf [hunk 2]
```diff
@@ -0,0 +1,44 @@
+resource "aws_waf_rate_based_rule" "BrokenAuthSessionManageRule" {
```
Comment (svaranasi-traderev) [inline WAFBrokenAuthSessionManageRule.tf]: Can you explain in a few words, or point me to documentation to, how this rule works?

File: kinesis_s3.tf [hunk 3]
```diff
@@ -0,0 +1,26 @@
+resource "aws_s3_bucket" "waf-logging-bucket" {
+  bucket = "${var.logging_bucket_name}"
+  acl = "log-delivery-write"
+
+  tags {
+    Name = "WAF Security Automations"
+    Enviroment = "nthgendevops"
```
Comment (svaranasi-traderev) [inline kinesis_s3.tf]: This tag, in all occurrences, should be a variable or dynamically fetched to allow deploying this solution to a different account.

File: kinesis_s3.tf [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  acl    = "private"
+
+  logging {
+    target_bucket = "${aws_s3_bucket.waf-logging-bucket.bucket}"
+    target_prefix = "s3logs"
+  }
+
+  tags {
+    Name = "WAF Security Automations"
+    Enviroment = "nthgendevops"
```
Comment (svaranasi-traderev) [inline kinesis_s3.tf]: This tag, in all occurrences, should be a variable or dynamically fetched to allow deploying this solution to a different account.

File: lambdawafbadbotparserfunction.tf [hunk 5]
```diff
@@ -42,4 +43,9 @@ resource "aws_lambda_function" "LambdaWAFBadBotParserFunction" {
             UUID = "${uuid()}"
         }
     }
+
+    tags {
+        Name = "WAF Security Automations"
+        Enviroment = "nthgendevops"
```
Comment (svaranasi-traderev) [inline lambdawafbadbotparserfunction.tf]: This tag, in all occurrences, should be a variable or dynamically fetched to allow deploying this solution to a different account.

File: lambdawaflogparserfunction.tf [hunk 6]
```diff
@@ -45,4 +46,9 @@ resource "aws_lambda_function" "LambdaWAFLogParserFunction" {
             UUID = "${uuid()}"
         }
     }
+
+    tags {
+        Name = "WAF Security Automations"
+        Enviroment = "nthgendevops"
```
Comment (svaranasi-traderev) [inline lambdawaflogparserfunction.tf]: This tag, in all occurrences, should be a variable or dynamically fetched to allow deploying this solution to a different account.

File: lambdawafreputationlistsparserfunction.tf [hunk 7]
```diff
@@ -40,4 +41,9 @@ resource "aws_lambda_function" "LambdaWAFReputationListsParserFunction" {
             SendAnonymousUsageData = "${var.SendAnonymousUsageData}"
         }
     }
+
+    tags{
+        Name = "WAF Security Automations"
+        Enviroment = "nthgendevops"
```
Comment (svaranasi-traderev) [inline lambdawafreputationlistsparserfunction.tf]: This tag, in all occurrences, should be a variable or dynamically fetched to allow deploying this solution to a different account.

File: solutionhelper.tf [hunk 8]
```diff
@@ -41,4 +42,9 @@ resource "aws_lambda_function" "SolutionHelper" {
             SendAnonymousUsageData = "${var.SendAnonymousUsageData}"
         }
     }
+
+    tags {
+        Name = "WAF Security Automations"
+        Enviroment = "nthgendevops"
```
Comment (svaranasi-traderev) [inline solutionhelper.tf]: This tag, in all occurrences, should be a variable or dynamically fetched to allow deploying this solution to a different account.

File: CloudfrontAccessLogS3Bucket.tf [hunk 9]
```diff
@@ -0,0 +1,9 @@
+resource "aws_s3_bucket" "WAFCloudfrontAccessLog" {
+  bucket = "${var.CloudFrontAccessLogBucket}"
+  acl = "log-delivery-write"
+
+  tags{
+    Name = "WAF Security Automations"
+    Enviroment = "nthgendevops"
```
Comment (svaranasi-traderev) [inline CloudfrontAccessLogS3Bucket.tf]: This tag, in all occurrences, should be a variable or dynamically fetched to allow deploying this solution to a different account.

File: lambdawafcustomresourcefunction.tf [hunk 10]
```diff
@@ -40,5 +41,10 @@ resource "aws_lambda_function" "LambdaWAFCustomResourceFunction" {
             SendAnonymousUsageData = "${var.SendAnonymousUsageData}"
         }
     }
+
+    tags {
+        Name = "WAF Security Automations"
+        Enviroment = "nthgendevops"
```
Comment (svaranasi-traderev) [inline lambdawafcustomresourcefunction.tf]: This tag, in all occurrences, should be a variable or dynamically fetched to allow deploying this solution to a different account.

File: solutionhelper.tf [hunk 11]
```diff
@@ -15,15 +15,16 @@
 ###############################################################################
 
 resource "aws_lambda_function" "SolutionHelper" {
-    depends_on = ["aws_s3_bucket_object.SolutionHelperZip"]
-    function_name = "${var.customer}-SolutionHelper-${element(split("-",uuid()),0)}"
+    #depends_on = ["aws_s3_bucket_object.SolutionHelperZip"]
```
Comment (svaranasi-traderev) [inline solutionhelper.tf]: Remove commented code.

File: misc/.template.main.tf [hunk 12]
```diff
@@ -38,11 +38,11 @@ variable "CloudFrontAccessLogBucket" {
 ##################################
 #default = "50"
```
Comment (svaranasi-traderev) [inline misc/.template.main.tf]: Remove commented code.

File: lambdawafreputationlistsparserfunction.tf [hunk 13]
```diff
@@ -15,14 +15,15 @@
 ###############################################################################
 
 resource "aws_lambda_function" "LambdaWAFReputationListsParserFunction" {
-    depends_on = ["aws_s3_bucket_object.ReputationListsParserZip"]
-    function_name = "${var.customer}-LambdaWAFReputationListsParserFunction-${element(split("-",uuid()),0)}"
+    #depends_on = ["aws_s3_bucket_object.ReputationListsParserZip"]
```
Comment (svaranasi-traderev) [inline lambdawafreputationlistsparserfunction.tf]: Remove commented code.

File: lambdawaflogparserfunction.tf [hunk 14]
```diff
@@ -15,15 +15,16 @@
 ###############################################################################
 
 resource "aws_lambda_function" "LambdaWAFLogParserFunction" {
-    depends_on = ["aws_s3_bucket_object.LogParserZip"]
-    function_name = "${var.customer}-LambdaWAFLogParserFunction-${element(split("-",uuid()),0)}"
+    #depends_on = ["aws_s3_bucket_object.LogParserZip"]
```
Comment (svaranasi-traderev) [inline lambdawaflogparserfunction.tf]: Remove commented code.

File: lambdawafcustomresourcefunction.tf [hunk 15]
```diff
@@ -15,13 +15,14 @@
 ###############################################################################
 
 resource "aws_lambda_function" "LambdaWAFCustomResourceFunction" {
-    depends_on = ["aws_s3_bucket_object.CustomResourceZip"]
-    function_name = "${var.customer}-LambdaWAFCustomResourceFunction-${element(split("-",uuid()),0)}"
+    #depends_on = ["aws_s3_bucket_object.CustomResourceZip"]
```
Comment (svaranasi-traderev) [inline lambdawafcustomresourcefunction.tf]: Remove commented code.

File: kinesisfirehose.tf [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+  provisioner "local-exec" {
+    command = "aws waf put-logging-configuration --logging-configuration 'ResourceArn=arn:aws:waf::${var.aws-account-id}:webacl/${aws_waf_web_acl.WAFWebACL.id},LogDestinationConfigs=${aws_kinesis_firehose_delivery_stream.aws-waf-logs.arn},RedactedFields=${var.kinesis-log-config-redacted-list}'"
+  }
+
+
+  /*
+  provisioner "local-exec" {
+    command = "aws waf put-logging-configuration --logging-configuration ${file("kinesislogging_config.json")}"
+  }
+*/
```
Comment (svaranasi-traderev) [inline kinesisfirehose.tf]: Remove commented code.

File: variables.tf [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+variable "splunk_hec_endpoint" {
+  type = "string"
+  default = "https://http-inputs-karauctionservices.splunkcloud.com/services/collector/event"
+}
+
+variable "splunk_hec_token" {
+  type = "string"
+  default = "D1ABC097-FC14-4988-9E07-E7CEAB878F22"
```
[the remark this answers is not in the document]
Comment (tr-eliu) [reply variables.tf]: forgot to take that out

**Comment by svaranasi-traderev** [comment]
I believe you missed quite a few comments.

File: wafwebacl.tf (hunk 1 above)
Comment (tr-eliu) [reply wafwebacl.tf]: I don't remember that discussion either, there's no way to create a customize metric in terraform. In order to setup cloudwatch alarm for each rule, a metric must already be associated with  the rule.

File: WAFBrokenAuthSessionManageRule.tf (hunk 2 above)
Comment (tr-eliu) [reply WAFBrokenAuthSessionManageRule.tf]: The rules is based on a document that AWS made on the top 10 OWASP vulnerabilities, its on page 5 of the document.

File: CloudfrontAccessLogS3Bucket.tf [hunk 18]
```diff
@@ -4,6 +4,7 @@ resource "aws_s3_bucket" "WAFCloudfrontAccessLog" {
 
   tags{
     Name = "WAF Security Automations"
-    Enviroment = "nthgendevops"
+    Region = "${data.aws_region.current_region.name}"
+    Application = "${var.application}"
   }
 }
```
Comment (svaranasi-traderev) [inline CloudfrontAccessLogS3Bucket.tf]: Add newline please.

File: main.tf [hunk 19]
```diff
@@ -91,4 +89,10 @@ output "account_id" {
...
 }
 
-
+###############################################################################
+#Application
+###############################################################################
+variable "application" {
+  description = "The software use to provision the infrastructure"
+  default = "Terraform"
+}
```
Comment (svaranasi-traderev) [inline main.tf]: Add newline please.

**Pull request merged**